File: 09 oc 代码知识/MajqPageView/MajqPageView/ViewController.m
//
//  ViewController.m
//  MajqPageView
//
//  Created by JQ on 2017/9/16.
//  Copyright © 2017年 Majq. All rights reserved.
//

#import "ViewController.h"
#import "FirstViewController.h"
#import "SecondViewController.h"
#import "ThirdViewController.h"
#import "MajqTableView.h"
#import "MajqPageScrollView.h"
#define majqScreenW  [UIScreen mainScreen].bounds.size.width
#define majqScreenH  [UIScreen mainScreen].bounds.size.height

@interface ViewController ()<UITableViewDelegate,UITableViewDataSource,UIScrollViewDelegate>

@property(nonatomic,strong)MajqTableView *tableView;
@property(nonatomic,strong)UIView *headerView;

@property(nonatomic,strong)MajqPageScrollView *pageScrollView;
@property(nonatomic,strong)NSMutableArray *childVCs;

@property (nonatomic, assign) BOOL canScroll;
@property (nonatomic, assign) BOOL isTopIsCanNotMoveTabView;
@property (nonatomic, assign) BOOL isTopIsCanNotMoveTabViewPre;

@end

@implementation ViewController

- (NSMutableArray *)childVCs
{
    if (_childVCs == nil) {
        _childVCs = [[NSMutableArray array]init];
    }
    return _childVCs;
}

- (UITableView *)tableView
{
    if (_tableView == nil) {
        _tableView = [[MajqTableView alloc]initWithFrame:CGRectMake(0, 0, majqScreenW, majqScreenH) style:UITableViewStylePlain];
        _tableView.delegate = self;
        _tableView.dataSource = self;
        _tableView.showsVerticalScrollIndicator = NO;

        
    }
    return _tableView;
}

- (UIView *)headerView
{
    if (_headerView == nil) {
        _headerView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, majqScreenW, 200)];
        _headerView.backgroundColor = [UIColor orangeColor];
    }
    
    return _headerView;
}

- (MajqPageScrollView *)pageScrollView
{
    if (_pageScrollView == nil) {
        _pageScrollView = [[MajqPageScrollView alloc]initWithFrame:CGRectMake(0, 45, majqScreenW, majqScreenH - 45)];
        FirstViewController *vc1 = [[FirstViewController alloc]initWithCurrent:@"1"];
        SecondViewController *vc2 = [[SecondViewController alloc]init];
        ThirdViewController *vc3 = [[ThirdViewController alloc]init];
        vc1.view.tag = 1;
        vc2.view.tag = 1;
        vc3.view.tag = 1;

        
        [self.childVCs addObject:vc1];
        [self.childVCs addObject:vc2];
        [self.childVCs addObject:vc3];
        
       
        

        [self.pageScrollView addController:self.childVCs];
        
        [self addChildViewController:vc1];
        [self addChildViewController:vc2];
        [self addChildViewController:vc3];
        [self.pageScrollView setCurrentIndex:0];
        
        
        
    }
    return _pageScrollView;

}

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    [self setUpConfig];

    
}

- (void)setUpConfig {
    [self.view addSubview: self.tableView];
    self.tableView.tableHeaderView = self.headerView;
    self.tableView.rowHeight = majqScreenH - 45;
}

#pragma mark - 协议
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return 1;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    static NSString *cellId = @"cellId";
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:cellId];
    if (cell == nil) {
        cell = [[UITableViewCell alloc]initWithStyle:UITableViewCellStyleDefault reuseIdentifier:cellId];
        cell.backgroundColor = [UIColor redColor];
        [cell.contentView addSubview:self.pageScrollView];
        cell.selectionStyle = UITableViewCellSelectionStyleNone;
    }
    
    return cell;
}


- (void)scrollViewDidScroll:(UIScrollView *)scrollView
{
    

}




@end

File: 02 Swift _tip/swift 3.0 练习知识点/UseMethod/UseMethod/ViewController.swift
//
//  ViewController.swift
//  UseMethod
//
//  Created by JQ on 2017/9/8.
//  Copyright © 2017年 Majq. All rights reserved.
//

import UIKit

class Person {
    var name: String = "11"
}




class ViewController: UIViewController {

    var _tableView: MajqTableView = MajqTableView(frame: .zero, style: .plain)
    let cellID = "id"
    let h = UIScreen.main.bounds.size.height
    override func viewDidLoad() {
        super.viewDidLoad()
        
       test02()
        
        
    }
    
    
    private func test01() {
    
        let v =  TestView(frame: CGRect(x: 20, y: 20, width: 20, height: 20))
        v.backgroundColor = UIColor.orange
        print(v)

        for subv in view.subviews {
            print(" class = \(type(of: subv))")
        }
        view.addSubview(v)
        
        view.addSubview(v)
        view.addSubview(v)
        view.addSubview(v)
        view.addSubview(v)
        for subv in view.subviews {
            print(" class = \(type(of: subv))")
        }


        
    }
    
    private func test02() {
    
        let v =  TestView(frame: CGRect(x: 20, y: 20, width: 20, height: 20))
        v.backgroundColor = UIColor.orange
        print("v 的信息\(v)")
        
        
        let v2 =  MView(frame: view.bounds)
        view.addSubview(v2)
        
        
        _tableView.delegate = self
        _tableView.dataSource = self
        _tableView.frame = CGRect(x: 20, y: 0, width: 375, height: h)
        _tableView.register(UITableViewCell.self, forCellReuseIdentifier: cellID)
        _tableView.rowHeight = 80
        
        v2.addSubview(_tableView)

        testTableHeaderView_Y(withView: v)

        //testTableHeaderView_N(withView: v)
        
    }
    
    private func testTableHeaderView_Y(withView v: TestView) {
        _tableView.tableHeaderView = v
        for subv in _tableView.subviews {
            print(" class = \(type(of: subv))")
        }
        
        print("----1----")
        print(_tableView)
        
        print("----2----")

        let c1 = _tableView.tableHeaderView?.superview
        print(c1!)
        print(NSStringFromClass(type(of: c1!)))
        print(" class = \(type(of: c1))")
        
        print("----3----")

        let c2 = _tableView.superview
        print(c2!)
        print(" class = \(type(of: c2))")

        /*
         Optional(<UseMethod.TestView: 0x7fb176d06830; frame = (0 0; 375 20); layer = <CALayer: 0x60000003b180>>)
         Optional(<UITableView: 0x7fb17702d400; frame = (20 0; 375 736); clipsToBounds = YES; gestureRecognizers = <NSArray: 0x600000053ad0>; layer = <CALayer: 0x60000003b520>; contentOffset: {0, 0}; contentSize: {375, 1620}>)

         */
    }
    
    private func testTableHeaderView_N(withView v: TestView) {
        
        for subv in _tableView.subviews {
            print(" class = \(type(of: subv))")
        }
        
        print(_tableView.tableHeaderView)
        print(_tableView.tableHeaderView?.superview)
        
        /*
         nil
         nil
         */
    }
    
  
    

}

extension ViewController: UITableViewDelegate,UITableViewDataSource {
    
    

    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return 20
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: cellID, for: indexPath)
        cell.textLabel?.text = "\(indexPath.row)"
        cell.contentView.backgroundColor = UIColor.gray

        return cell
    }
}

